mio/io.py

```diff
@@ -160,26 +163,33 @@ class BufferedCSVWriter:
         The buffer for storing rows before writing.
     """
 
-    def __init__(self, file_path: Union[str, Path], buffer_size: int = 100):
+    def __init__(self, file_path: Union[str, Path], header: list[str], buffer_size: int = 100):
         self.file_path: Path = Path(file_path)
+        self.header = header
         self.buffer_size = buffer_size
         self.buffer = []
         self.logger = init_logger("BufferedCSVWriter")
 
+        # write header in first row
+        self.buffer.append(self.header)
+
         # Ensure the buffer is flushed when the program exits
         atexit.register(self.flush_buffer)
 
-    def append(self, data: List[Any]) -> None:
+    def append(self, data: dict) -> None:
         """
         Append data (as a list) to the buffer.
 
         Parameters
         ----------
-        data : List[Any]
+        data : dict
             The data to be appended.
+            Rows are constructed and columns are ordered according to `header` -
+            keys that are not in `header` are ignored, and missing keys are `None`
         """
-        data = [int(value) if isinstance(value, np.generic) else value for value in data]
-        self.buffer.append(data)
+        row = [data.get(key) for key in self.header]
+        row = [int(value) if isinstance(value, np.generic) else value for value in row]
```

Comment:
This conversion looks unnecessary, as the list always comes from a pydantic model now (with native Python number types, not `np` types), and anyway, casting `np.generic` to `int` seems extremely broad.

I think I initially added this conversion to handle `np.uintX`, but I guess this should be handled in the model.

```suggestion
```

Reply:
i *believe* we use `model_construct` for the headers since they are in a perf-sensitive loop, so in this case i think we do need the int coercion, but we should test this in any case. could be the case that at some point all the coercion outside the pydantic validation becomes more expensive than just using the validation directly, seems worth profiling.

Reply:
Ah, thanks, I overlooked the `construct=True` in `_parse_header` because this part isn't as perf-sensitive as the data receiving process (the header construction is in a process that grabs buffers from the queue populated by the hardware interfacing process).

I think it's fine to use Pydantic validation for this, as it's not in the first FPGA loop (maybe better to profile, though I'm not confident about doing it myself).

Reply:
This mostly felt confusing to me because this CSV writer also handles float values like UNIX time (we want sub-sec resolution).